File: problems/225/solution.ts
// Accepted: 54ms (59.04%), 42.84MB (55.42%)

class MyQueue {
    private queue: number[];
    private popIndex: number;
    private pushIndex: number;

    constructor() {
        this.queue = [];
        this.popIndex = 0;
        this.pushIndex = 0;
    }

    push(value: number): void {
        this.queue[this.pushIndex++] = value;
    }

    pop(): number {
        return this.queue[this.popIndex++];
    }

    peek(): number {
        return this.queue[this.popIndex];
    }

    size(): number {
        return this.pushIndex - this.popIndex;
    }
}

class MyStack {
    private queue: MyQueue;

    constructor() {
        this.queue = new MyQueue();
    }

    push(x: number): void {
        this.queue.push(x);
        for (let i = 0; i < this.queue.size()-1; ++i) {
            this.queue.push(this.queue.pop())
        }
    }

    pop(): number {
        return this.queue.pop();
    }

    top(): number {
        return this.queue.peek();
    }

    empty(): boolean {
        return this.queue.size() === 0;
    }
}

/**
 * Your MyStack object will be instantiated and called as such:
 * var obj = new MyStack()
 * obj.push(x)
 * var param_2 = obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.empty()
 */

File: problems/225/solution.cpp
// Accepted: 0ms (100.00%), 7.29MB (51.84%)

class MyQueue {
    queue<int> queue;
public:
    void push(int value) {
        queue.push(value);
    }

    int pop() {
        int pop = queue.front();
        queue.pop();
        return pop;
    }

    int peek() {
        return queue.front();
    }

    int size() {
        return queue.size();
    }
};

class MyStack {
    MyQueue queue;
public:
    void push(int x) {
        queue.push(x);
        for (int i = 0; i < queue.size()-1; ++i) {
            queue.push(queue.pop());
        }
    }

    int pop() {
        return queue.pop();
    }

    int top() {
        return queue.peek();
    }

    bool empty() {
        return queue.size() == 0;
    }
};

/**
 * Your MyStack object will be instantiated and called as such:
 * MyStack* obj = new MyStack();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->top();
 * bool param_4 = obj->empty();
 */
